File: src/main/java/org/example/kalkulationsprogramm/controller/FirmaController.java
package org.example.kalkulationsprogramm.controller;

import lombok.RequiredArgsConstructor;
import org.example.kalkulationsprogramm.dto.EmailAbsenderDto;
import org.example.kalkulationsprogramm.dto.FirmeninformationDto;
import org.example.kalkulationsprogramm.dto.KostenstelleDto;
import org.example.kalkulationsprogramm.dto.SteuerberaterKontaktDto;
import org.example.kalkulationsprogramm.domain.KostenstellenTyp;
import org.example.kalkulationsprogramm.service.EmailAbsenderService;
import org.example.kalkulationsprogramm.service.FirmeninformationService;
import org.example.kalkulationsprogramm.service.KostenstelleService;
import org.example.kalkulationsprogramm.service.SteuerberaterKontaktService;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.List;
import java.util.Map;

/**
 * REST Controller für Firmenstammdaten, Kostenstellen und Steuerberater.
 */
@RestController
@RequestMapping("/api/firma")
@RequiredArgsConstructor
public class FirmaController {

    private final FirmeninformationService firmeninformationService;
    private final KostenstelleService kostenstelleService;
    private final SteuerberaterKontaktService steuerberaterKontaktService;
    private final EmailAbsenderService emailAbsenderService;

    // ==================== FIRMENINFORMATION ====================

    @GetMapping
    public ResponseEntity<FirmeninformationDto> getFirmeninformation() {
        return ResponseEntity.ok(firmeninformationService.getFirmeninformation());
    }

    @PutMapping
    public ResponseEntity<FirmeninformationDto> speichernFirmeninformation(@RequestBody FirmeninformationDto dto) {
        return ResponseEntity.ok(firmeninformationService.speichern(dto));
    }

    // ==================== FIRMENLOGO ====================

    /**
     * Laedt das Firmenlogo hoch (multipart). Whitelist: PNG, JPEG, WebP.
     * Der Client-Dateiname wird bewusst ignoriert – wir speichern unter
     * {@code logo.<ext>}, um Pfad-Traversal komplett auszuschliessen.
     */
    @PostMapping(value = "/logo", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public ResponseEntity<?> uploadLogo(@RequestParam("datei") MultipartFile datei) {
        try {
            FirmeninformationDto dto = firmeninformationService.speichereLogoDatei(datei);
            return ResponseEntity.ok(dto);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(Map.of("message", e.getMessage()));
        } catch (IOException e) {
            return ResponseEntity.internalServerError().body(Map.of("message", "Speichern fehlgeschlagen"));
        }
    }

    /**
     * Liefert das aktuell hinterlegte Firmenlogo als Binary-Response.
     * 404, falls kein Logo gepflegt ist.
     */
    @GetMapping("/logo")
    public ResponseEntity<byte[]> downloadLogo() {
        byte[] bytes = firmeninformationService.loadLogoBytes();
        if (bytes == null) {
            return ResponseEntity.notFound().build();
        }
        String contentType = firmeninformationService.ermittleLogoContentType();
        MediaType mediaType = contentType != null
                ? MediaType.parseMediaType(contentType)
                : MediaType.APPLICATION_OCTET_STREAM;
        return ResponseEntity.ok().contentType(mediaType).body(bytes);
    }

    @DeleteMapping("/logo")
    public ResponseEntity<FirmeninformationDto> deleteLogo() {
        return ResponseEntity.ok(firmeninformationService.loescheLogoDatei());
    }

    // ==================== KOSTENSTELLEN ====================

    @GetMapping("/kostenstellen")
    public ResponseEntity<List<KostenstelleDto>> getKostenstellen() {
        return ResponseEntity.ok(kostenstelleService.findAll());
    }

    @GetMapping("/kostenstellen/typ/{typ}")
    public ResponseEntity<List<KostenstelleDto>> getKostenstellenByTyp(@PathVariable KostenstellenTyp typ) {
        return ResponseEntity.ok(kostenstelleService.findByTyp(typ));
    }

    @GetMapping("/kostenstellen/{id}")
    public ResponseEntity<KostenstelleDto> getKostenstelle(@PathVariable Long id) {
        KostenstelleDto dto = kostenstelleService.findById(id);
        if (dto == null) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(dto);
    }

    @PostMapping("/kostenstellen")
    public ResponseEntity<KostenstelleDto> erstelleKostenstelle(@RequestBody KostenstelleDto dto) {
        dto.setId(null); // Neue Kostenstelle
        return ResponseEntity.ok(kostenstelleService.speichern(dto));
    }

    @PutMapping("/kostenstellen/{id}")
    public ResponseEntity<KostenstelleDto> aktualisiereKostenstelle(
            @PathVariable Long id, 
            @RequestBody KostenstelleDto dto) {
        dto.setId(id);
        return ResponseEntity.ok(kostenstelleService.speichern(dto));
    }

    @DeleteMapping("/kostenstellen/{id}")
    public ResponseEntity<Void> loescheKostenstelle(@PathVariable Long id) {
        kostenstelleService.loeschen(id);
        return ResponseEntity.noContent().build();
    }

    @PostMapping("/kostenstellen/init")
    public ResponseEntity<List<KostenstelleDto>> initKostenstellen() {
        kostenstelleService.erstelleStandardKostenstellen();
        return ResponseEntity.ok(kostenstelleService.findAll());
    }

    // ==================== STEUERBERATER ====================

    @GetMapping("/steuerberater")
    public ResponseEntity<List<SteuerberaterKontaktDto>> getSteuerberater() {
        return ResponseEntity.ok(steuerberaterKontaktService.findAll());
    }

    @GetMapping("/steuerberater/{id}")
    public ResponseEntity<SteuerberaterKontaktDto> getSteuerberaterById(@PathVariable Long id) {
        SteuerberaterKontaktDto dto = steuerberaterKontaktService.findById(id);
        if (dto == null) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(dto);
    }

    @PostMapping("/steuerberater")
    public ResponseEntity<SteuerberaterKontaktDto> erstelleSteuerberater(@RequestBody SteuerberaterKontaktDto dto) {
        dto.setId(null);
        return ResponseEntity.ok(steuerberaterKontaktService.speichern(dto));
    }

    @PutMapping("/steuerberater/{id}")
    public ResponseEntity<SteuerberaterKontaktDto> aktualisiereSteuerberater(
            @PathVariable Long id, 
            @RequestBody SteuerberaterKontaktDto dto) {
        dto.setId(id);
        return ResponseEntity.ok(steuerberaterKontaktService.speichern(dto));
    }

    @DeleteMapping("/steuerberater/{id}")
    public ResponseEntity<Void> loescheSteuerberater(@PathVariable Long id) {
        steuerberaterKontaktService.loeschen(id);
        return ResponseEntity.noContent().build();
    }

    // ==================== E-MAIL-ABSENDER ====================

    @GetMapping("/email-absender")
    public ResponseEntity<List<EmailAbsenderDto>> getEmailAbsender() {
        return ResponseEntity.ok(emailAbsenderService.findAll());
    }

    @PostMapping("/email-absender")
    public ResponseEntity<?> erstelleEmailAbsender(@RequestBody EmailAbsenderDto dto) {
        try {
            dto.setId(null);
            return ResponseEntity.ok(emailAbsenderService.save(dto));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.badRequest().body(Map.of("message", ex.getMessage()));
        }
    }

    @PutMapping("/email-absender/{id}")
    public ResponseEntity<?> aktualisiereEmailAbsender(
            @PathVariable Long id,
            @RequestBody EmailAbsenderDto dto) {
        try {
            dto.setId(id);
            return ResponseEntity.ok(emailAbsenderService.save(dto));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.badRequest().body(Map.of("message", ex.getMessage()));
        }
    }

    @DeleteMapping("/email-absender/{id}")
    public ResponseEntity<Void> loescheEmailAbsender(@PathVariable Long id) {
        emailAbsenderService.delete(id);
        return ResponseEntity.noContent().build();
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/controller/FirmaController.kt
package org.example.kalkulationsprogramm.controller

import org.example.kalkulationsprogramm.domain.KostenstellenTyp
import org.example.kalkulationsprogramm.dto.EmailAbsenderDto
import org.example.kalkulationsprogramm.dto.FirmeninformationDto
import org.example.kalkulationsprogramm.dto.KostenstelleDto
import org.example.kalkulationsprogramm.dto.SteuerberaterKontaktDto
import org.example.kalkulationsprogramm.service.EmailAbsenderService
import org.example.kalkulationsprogramm.service.FirmeninformationService
import org.example.kalkulationsprogramm.service.KostenstelleService
import org.example.kalkulationsprogramm.service.SteuerberaterKontaktService
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.multipart.MultipartFile
import java.io.IOException

@RestController
@RequestMapping("/api/firma")
class FirmaController(
    private val firmeninformationService: FirmeninformationService,
    private val kostenstelleService: KostenstelleService,
    private val steuerberaterKontaktService: SteuerberaterKontaktService,
    private val emailAbsenderService: EmailAbsenderService,
) {
    @GetMapping
    fun getFirmeninformation(): ResponseEntity<FirmeninformationDto> {
        return ResponseEntity.ok(firmeninformationService.firmeninformation)
    }

    @PutMapping
    fun speichernFirmeninformation(@RequestBody dto: FirmeninformationDto): ResponseEntity<FirmeninformationDto> {
        return ResponseEntity.ok(firmeninformationService.speichern(dto))
    }

    @PostMapping(value = ["/logo"], consumes = [MediaType.MULTIPART_FORM_DATA_VALUE])
    fun uploadLogo(@RequestParam("datei") datei: MultipartFile): ResponseEntity<Any> {
        return try {
            ResponseEntity.ok(firmeninformationService.speichereLogoDatei(datei))
        } catch (e: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("message" to e.message))
        } catch (e: IOException) {
            ResponseEntity.internalServerError().body(mapOf("message" to "Speichern fehlgeschlagen"))
        }
    }

    @GetMapping("/logo")
    fun downloadLogo(): ResponseEntity<ByteArray> {
        val bytes = firmeninformationService.loadLogoBytes() ?: return ResponseEntity.notFound().build()
        val contentType = firmeninformationService.ermittleLogoContentType()
        val mediaType = if (contentType != null) MediaType.parseMediaType(contentType) else MediaType.APPLICATION_OCTET_STREAM
        return ResponseEntity.ok().contentType(mediaType).body(bytes)
    }

    @DeleteMapping("/logo")
    fun deleteLogo(): ResponseEntity<FirmeninformationDto> {
        return ResponseEntity.ok(firmeninformationService.loescheLogoDatei())
    }

    @GetMapping("/kostenstellen")
    fun getKostenstellen(): ResponseEntity<List<KostenstelleDto>> {
        return ResponseEntity.ok(kostenstelleService.findAll())
    }

    @GetMapping("/kostenstellen/typ/{typ}")
    fun getKostenstellenByTyp(@PathVariable typ: KostenstellenTyp): ResponseEntity<List<KostenstelleDto>> {
        return ResponseEntity.ok(kostenstelleService.findByTyp(typ))
    }

    @GetMapping("/kostenstellen/{id}")
    fun getKostenstelle(@PathVariable id: Long): ResponseEntity<KostenstelleDto> {
        val dto = kostenstelleService.findById(id) ?: return ResponseEntity.notFound().build()
        return ResponseEntity.ok(dto)
    }

    @PostMapping("/kostenstellen")
    fun erstelleKostenstelle(@RequestBody dto: KostenstelleDto): ResponseEntity<KostenstelleDto> {
        dto.id = null
        return ResponseEntity.ok(kostenstelleService.speichern(dto))
    }

    @PutMapping("/kostenstellen/{id}")
    fun aktualisiereKostenstelle(@PathVariable id: Long, @RequestBody dto: KostenstelleDto): ResponseEntity<KostenstelleDto> {
        dto.id = id
        return ResponseEntity.ok(kostenstelleService.speichern(dto))
    }

    @DeleteMapping("/kostenstellen/{id}")
    fun loescheKostenstelle(@PathVariable id: Long): ResponseEntity<Void> {
        kostenstelleService.loeschen(id)
        return ResponseEntity.noContent().build()
    }

    @PostMapping("/kostenstellen/init")
    fun initKostenstellen(): ResponseEntity<List<KostenstelleDto>> {
        kostenstelleService.erstelleStandardKostenstellen()
        return ResponseEntity.ok(kostenstelleService.findAll())
    }

    @GetMapping("/steuerberater")
    fun getSteuerberater(): ResponseEntity<List<SteuerberaterKontaktDto>> {
        return ResponseEntity.ok(steuerberaterKontaktService.findAll())
    }

    @GetMapping("/steuerberater/{id}")
    fun getSteuerberaterById(@PathVariable id: Long): ResponseEntity<SteuerberaterKontaktDto> {
        val dto = steuerberaterKontaktService.findById(id) ?: return ResponseEntity.notFound().build()
        return ResponseEntity.ok(dto)
    }

    @PostMapping("/steuerberater")
    fun erstelleSteuerberater(@RequestBody dto: SteuerberaterKontaktDto): ResponseEntity<SteuerberaterKontaktDto> {
        dto.id = null
        return ResponseEntity.ok(steuerberaterKontaktService.speichern(dto))
    }

    @PutMapping("/steuerberater/{id}")
    fun aktualisiereSteuerberater(
        @PathVariable id: Long,
        @RequestBody dto: SteuerberaterKontaktDto,
    ): ResponseEntity<SteuerberaterKontaktDto> {
        dto.id = id
        return ResponseEntity.ok(steuerberaterKontaktService.speichern(dto))
    }

    @DeleteMapping("/steuerberater/{id}")
    fun loescheSteuerberater(@PathVariable id: Long): ResponseEntity<Void> {
        steuerberaterKontaktService.loeschen(id)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/email-absender")
    fun getEmailAbsender(): ResponseEntity<List<EmailAbsenderDto>> {
        return ResponseEntity.ok(emailAbsenderService.findAll())
    }

    @PostMapping("/email-absender")
    fun erstelleEmailAbsender(@RequestBody dto: EmailAbsenderDto): ResponseEntity<Any> {
        return try {
            dto.id = null
            ResponseEntity.ok(emailAbsenderService.save(dto))
        } catch (ex: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("message" to ex.message))
        }
    }

    @PutMapping("/email-absender/{id}")
    fun aktualisiereEmailAbsender(@PathVariable id: Long, @RequestBody dto: EmailAbsenderDto): ResponseEntity<Any> {
        return try {
            dto.id = id
            ResponseEntity.ok(emailAbsenderService.save(dto))
        } catch (ex: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("message" to ex.message))
        }
    }

    @DeleteMapping("/email-absender/{id}")
    fun loescheEmailAbsender(@PathVariable id: Long): ResponseEntity<Void> {
        emailAbsenderService.delete(id)
        return ResponseEntity.noContent().build()
    }
}
